src/common/framework/params.ts

```diff
@@ -1,112 +1,131 @@
-import {
-  ParamArgument,
-  ParamSpec,
-  ParamSpecIterable,
-  ParamSpecIterator,
-  paramsEquals,
-} from './params_utils.js';
+import { ParamSpec, ParamSpecIterable, paramsEquals } from './params_utils.js';
 import { assert } from './util/util.js';
 
-export function poptions(name: string, values: ParamArgument[]): POptions {
-  return new POptions(name, values);
-}
-export function pbool(name: string): POptions {
-  return new POptions(name, [false, true]);
-}
-export function pexclude(params: ParamSpecIterable, exclude: ParamSpecIterable): PExclude {
-  return new PExclude(params, exclude);
+export function poptions<Name extends string, V>(
+  name: Name,
+  values: Iterable<V>
+): string extends Name ? never : Iterable<{ [name in Name]: V }> {
+  return reusableGenerator(function* () {
+    for (const value of values) {
+      yield { [name]: value };
+    }
+  }) as any;
 }
-export function pfilter(cases: ParamSpecIterable, pred: Predicate): PFilter {
-  return new PFilter(cases, pred);
+
+export function pbool<Name extends string>(
+  name: Name
+): string extends Name ? never : Iterable<{ [name in Name]: boolean }> {
+  return poptions(name, [false, true]);
 }
-export function pcombine(...params: ParamSpecIterable[]): ParamSpecIterable {
-  return new PCombine(params);
+
+export function params(): ParamsBuilder<{}> {
+  return new ParamsBuilder();
 }
 
-class POptions implements ParamSpecIterable {
-  private name: string;
-  private values: ParamArgument[];
+class ParamsBuilder<A extends {}> implements ParamSpecIterable {
+  private paramSpecs: ParamSpecIterable = [{}];
 
-  constructor(name: string, values: ParamArgument[]) {
-    this.name = name;
-    this.values = values;
+  [Symbol.iterator](): Iterator<A> {
+    const iter: Iterator<ParamSpec> = this.paramSpecs[Symbol.iterator]();
+    return iter as Iterator<A>;
   }
 
-  *[Symbol.iterator](): ParamSpecIterator {
-    for (const value of this.values) {
-      yield { [this.name]: value };
+  combine<B extends {}>(
+    newParams: Iterable<B>
+  ): ParamsBuilder<
+    {
+      [K in keyof A | keyof B]: K extends keyof A
+        ? K extends keyof B
+          ? never
+          : A[K]
+        : K extends keyof B
+        ? B[K]
+        : never;
     }
+  > {
```

Comment:
note this is the same in every place, but it's not typedeffed because it makes the inferred types unreadable.